tensorflow/contrib/opt/python/training/irprop_plus.py

```diff
@@ -0,0 +1,284 @@
+        state.get_hyper("delta_min", var.dtype.base_dtype),
+        state.get_hyper("delta_max", var.dtype.base_dtype),
+        math_ops.cast(update_error, var.dtype.base_dtype),
+        math_ops.cast(old_error, var.dtype.base_dtype),
+        grad, use_locking=self._use_locking)
+
+  def minimize(self, loss, global_step=None, var_list=None,
+               gate_gradients=optimizer_v2.OptimizerV2.GATE_OP,
+               aggregation_method=None, colocate_gradients_with_ops=False,
+               name=None, grad_loss=None):
+    """Add operations to minimize `loss` by updating `var_list`.
+    This method simply combines calls `compute_gradients()` and
+    `apply_gradients()`. If you want to process the gradient before applying
+    them call `compute_gradients()` and `apply_gradients()` explicitly instead
+    of using this function. The loss argument has to be passed as a 0-D Tensor.
+    Args:
+      loss: A 0-D `Tensor` containing the value to minimize.
+      global_step: Optional `Variable` to increment by one after the
+        variables have been updated.
+      var_list: Optional list or tuple of `Variable` objects to update to
+        minimize `loss`.  Defaults to the list of variables collected in
+        the graph under the key `GraphKeys.TRAINABLE_VARIABLES`.
+      gate_gradients: How to gate the computation of gradients.  Can be
+        `GATE_NONE`, `GATE_OP`, or  `GATE_GRAPH`.
+      aggregation_method: Specifies the method used to combine gradient terms.
+        Valid values are defined in the class `AggregationMethod`.
+      colocate_gradients_with_ops: If True, try colocating gradients with
+        the corresponding op.
+      name: Optional name for the returned operation.
+      grad_loss: Optional. A `Tensor` holding the gradient computed for `loss`.
+    Returns:
+      An Operation that updates the variables in `var_list`.  If `global_step`
+      was not `None`, that operation also increments `global_step`.
+    Raises:
+      ValueError: If some of the variables are not `Variable` objects.
+      ValueError: If the `loss` is not a 0-D tensor (scalar).
+    @compatibility(eager)
+    When eager execution is enabled, `loss` should be a Python function that
+    takes elements of `var_list` as arguments and computes the value to be
+    minimized. If `var_list` is `None`, `loss` should take no arguments.
+    Minimization (and gradient computation) is done with respect to the
+    elements of `var_list` if not `None`, else with respect to any trainable
+    variables created during the execution of the `loss` function.
+    `gate_gradients`, `aggregation_method`, `colocate_gradients_with_ops` and
+    `grad_loss` are ignored when eager execution is enabled.
+    @end_compatibility
+    """
+    # Override method from base class, the loss is required to be scalar
+
+    # Error E(t)
+    if not self._is_scalar(loss):
+      raise ValueError("'loss' (%s) must be a 0-D tensor." % loss)
```

Comment:
Can you drop these scalar tests? I think they are redundant / not present in other optimizers. TF will silently reduce non-scalar losses to scalars I think. If you drop this you can stop overriding apply_gradients and minimize and your code will be shorter and easier to read and more future-proof.

Reply:
I've tested this part and TF does not reduce the loss to scalar. Although, not providing a reduce op for the loss does work for other optimizers, for iRprop+ it is required.
An alternative would be to provide a default reduce operation in case none is provided. What do you think?

Reply:
tf.gradients implicitly reduces the loss to a scalar so you do not need this.

Reply:
@alextp It does not reduce the loss to a scalar for me (tested in OptimizerV2), unless the weights are scalars. For example, after running the code below, the cost is a vector which needs to be reduced to a scalar in order to feed the error to IRprop+.  Am I missing something? some help would be much appreciated. Thank you.

import tensorflow as tf

x1 = tf.Variable([1.0, 2.0])
x2 = tf.Variable([2.0, 3.0])
loss = 2*x1**2 + 3*x2

grad = tf.gradients(loss, [x1, x2])

init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    grad_value = sess.run(grad)
    print(sess.run(loss))